### PR TITLE
Chitter-challenge - WK4

### PR DESCRIPTION
**Pull request by sandyMax974**
All `STRAIGHT UP` user stories have been implemented. I didn't get a chance to attempt the harder one as I decided to play with CSS instead.
I didn't get to implements solution for some edge cases but I've put my thoughts at the bottom of the README.

Feedbacks on any of the below would be welcome:

- Use of isolated test
- Use of SRP
- Use of RESTful route
- How to handle very long database query string
- Anything else

### CODE REVIEW

File: lib/chitter.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+require_relative 'peep'
+
+class Chitter
+
+  def self.create(username, content)
+    connection = PG.connect :dbname => "chitter_#{ENV['RACK_ENV']}"
+    result = connection.exec("SELECT * FROM users WHERE username = '#{username}'").first
+    insertion = connection.exec("INSERT INTO peeps (user_id, posted_time, content) VALUES(#{result['id']}, current_timestamp, '#{content}') RETURNING id, user_id, posted_time, content").first
```
Comment (kerisic) [inline lib/chitter.rb:8]: Could think about extracting database connection to run at the beginning of running the app - and extract connection.exec into a database connection class such as [this ](https://github.com/makersacademy/course/blob/master/bookmark_manager/walkthroughs/14.md)

File: views/feed.erb [hunk 2]
```diff
@@ -0,0 +1,81 @@
+<!DOCTYPE html>
+  <head>
+    <title>Home/Chitter</title>
+    <style>
+      body {
+        font-family: "Helvetica Neue", Roboto, "Segoe UI", Calibri, sans-serif;
+        background-color: white;
+        background-image: url("https://wallpaperaccess.com/full/1459046.png")
+      }
```
Comment (kerisic) [inline views/feed.erb:9]: Similar css across views can be saved in an external stylesheet where views could have a link to.

File: db/migration/02_seeting_up_peeps_database.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+CREATE TABLE peeps (id SERIAL PRIMARY KEY, user_id INT, posted_time TIMESTAMP, content VARCHAR(280));
```
Comment (kerisic) [inline db/migration/02_seeting_up_peeps_database.sql:1]: user_id in peeps could be referenced as a foreign key in table creation as it can be linked to the primary key in the user table

File: lib/peep.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  attr_reader :id, :username, :posted_time, :content
+
+  def initialize(id, username, posted_time, content)
+    @id = id
+    @posted_time = posted_time
+    @content = content
+    @username = username
+  end
+
+end
```
Comment (kerisic) [inline lib/peep.rb:11]: Good class encapsulation so that each class has a clear responsibility!

File: lib/user.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+class User
+  attr_reader :name, :lastname, :username, :email, :id
+
+  def initialize(name, lastname, username, email, password)
+    connection = PG.connect :dbname => "chitter_#{ENV['RACK_ENV']}"
+    insertion = connection.exec("INSERT INTO users (name, lastname, username, email, password) VALUES('#{name}', '#{lastname}', '#{username}', '#{email}', '#{password}') RETURNING id, name, lastname, username, email, password")
```
Comment (alicelieutier) [inline lib/user.rb:6]: It would be nicer to read this on several lines. You can look up ruby multiline strings as there are a number of solutions to this, and the one I give here may not be your favourite, but you could do:
```rb
query = %{
  INSERT INTO users (name, lastname, username, email, password)
  VALUES('#{name}', '#{lastname}', '#{username}', '#{email}', '#{password}')
  RETURNING id, name, lastname, username, email, password;
}
insertion = connection.exec(query)
```

Also this way you can easily log the query by adding a `puts query` between the query and insertion.

File: spec/features/feed_spec.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+feature 'Feed' do
+  scenario 'A user can see a feed of all posted peeps' do
+    connection = PG.connect :dbname => "chitter_#{ENV['RACK_ENV']}"
+    user = User.new('Sandy', 'Pounoussamy', 'sandy974', 'pounoussamy.sandy@gmail.com', '1234')
+    peep = "Who wants to know what I had for lunch?!"
+    connection.exec("INSERT INTO peeps (user_id, posted_time, content) VALUES('#{user.id}', current_timestamp, '#{peep}')")
```
Comment (alicelieutier) [inline spec/features/feed_spec.rb:6]: In a feature test, you may want to only interact with your app through the user interface. So you could have your user create tow peeps, and then see them, for example.

File: spec/features/post_peep_spec.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+feature 'Posting' do
+  scenario 'A user can post/create a new peep' do
+    visit('/signup')
+    signup
+
+    fill_in('content', with: "This is an automated peep")
+    click_button('Post')
+
+    expect(page).to have_content("This is an automated peep")
+  end
```
Comment (alicelieutier) [inline spec/features/post_peep_spec.rb:10]: This is good!

File: spec/units/chitter_spec.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
+require 'chitter'
+
+describe Chitter do
+
+  describe '.list' do
+    it 'display all the peeps' do
+      connection = PG.connect :dbname => "chitter_#{ENV['RACK_ENV']}"
+      demo_user_one = connection.exec("INSERT INTO users (name, lastname, username, email, password) VALUES('Hannah', 'Lowe', 'hanalo13', 'test@gmail.com', '1234') RETURNING id, name, lastname, username, email, password").first
+      demo_user_two = connection.exec("INSERT INTO users (name, lastname, username, email, password) VALUES('Sandy', 'Pounoussamy', 'sandy974', 'mytest@gmail.com', '1234') RETURNING id, name, lastname, username, email, password").first
```
Comment (alicelieutier) [inline spec/units/chitter_spec.rb:9]: These could be helper methods.

File: views/feed.erb (hunk 2 above)
Comment (alicelieutier) [reply views/feed.erb:9]: agreed.

**Review by alicelieutier — COMMENTED** [review]
Overall, your code looks like it works, so you've definitely cracked how to create a webapp that talks to a database.
What you can improve most on here IMO is the readability and organisation of your code. 
@kerisic added some good suggestions, and I would push even further, especially about naming and line length.

See comments below.

Here's an article I wrote about naming in code: https://medium.com/swlh/an-opinionated-guide-to-naming-your-code-aimed-at-new-developers-6f769d60e098
